app/src/block_hash.rs

```diff
@@ -1,37 +1,226 @@
 use {
-    alloy::primitives::B256, umi_blockchain::block::BlockQueries,
-    umi_evm_ext::state::BlockHashLookup,
+    alloy::primitives::B256,
+    std::{
+        collections::VecDeque,
+        sync::{Arc, RwLock},
+    },
+    umi_blockchain::block::BlockQueries,
+    umi_evm_ext::state::{BlockHashLookup, BlockHashWriter},
 };
 
-pub struct StorageBasedProvider<'a, S, B> {
-    storage: &'a S,
-    block_query: &'a B,
+const BLOCKHASH_HISTORY_SIZE: usize = 256;
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct BlockHashEntry {
+    number: u64,
+    hash: B256,
+}
+
+#[derive(Debug, Clone)]
+pub struct BlockHashRingBuffer {
+    entries: VecDeque<BlockHashEntry>,
+    /// Block number of the latest block stored for validation purposes
+    latest_block: u64,
 }
 
-impl<'a, S, B> StorageBasedProvider<'a, S, B>
-where
-    B: BlockQueries<Storage = S>,
-{
-    pub fn new(storage: &'a S, block_query: &'a B) -> Self {
+impl Default for BlockHashRingBuffer {
+    fn default() -> Self {
         Self {
-            storage,
-            block_query,
+            entries: VecDeque::with_capacity(BLOCKHASH_HISTORY_SIZE),
+            latest_block: 0,
         }
     }
 }
 
-// TODO: looking up the entire block just to get its hash is inefficient;
-// especially since we do not need the whole history, only the most recent 256
-// block hashes. This implementation should be replaced with something better.
-impl<S, B> BlockHashLookup for StorageBasedProvider<'_, S, B>
-where
-    B: BlockQueries<Storage = S>,
-{
+impl BlockHashRingBuffer {
+    pub fn push(&mut self, block_number: u64, block_hash: B256) {
+        // Ensure blocks are added in sequence (detect potential reorgs)
+        if block_number > 0 && self.latest_block > 0 && block_number != self.latest_block + 1 {
+            eprintln!(
```

Comment:
Should this stop the server?

Reply:
I feel like no, because the block hash functionality is just there for the speedup and nothing critical happens if the order is broken, though that's definitely an abnormality. I admit it would make sense to e.g. have some fallback to block queries as we were doing before.